特征工程主要分为三部分：
1. 数据预处理 对应的sklearn包：sklearn-Processing data
2. 特征选择 对应的sklearn包： sklearn-Feature selection
3. 降维 对应的sklearn包： sklearn-Dimensionality reduction

In [76]:
from sklearn.datasets import load_iris
import pandas as pd
import numpy as np

In [14]:
data = load_iris()

In [17]:
data.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [18]:
data.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [4]:
data.data.shape

(150, 4)

In [16]:
data.target.shape

(150,)

## 1.数据预处理

### 1.1无量纲化（数据规范化）

#### 1.1.1标准化（对列向量处理）

在机器学习中，我们可能要处理不同种类的资料，例如，音讯和图片上的像素值，这些资料可能是高维度的，资料标准化后会使每个特征中的数值**平均值μ变为0**(将每个特征的值都减掉原始资料中该特征的平均)、**标准差σ变为1**，这个方法被广泛的使用在许多机器学习算法中(例如：支持向量机、逻辑回归和类神经网络)。
$$x' = \frac{x - μ} {σ}$$

In [6]:
from sklearn.preprocessing import StandardScaler

In [7]:
std = StandardScaler()
std.fit_transform(data.data) 
'''
对数据标准化的目的是消除特征之间的差异性
不同特征对参数的影响程度就一样了
便于特征一心一意学习权重
'''

array([[-9.00681170e-01,  1.01900435e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00, -1.31979479e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.38535265e+00,  3.28414053e-01, -1.39706395e+00,
        -1.31544430e+00],
       [-1.50652052e+00,  9.82172869e-02, -1.28338910e+00,
        -1.31544430e+00],
       [-1.02184904e+00,  1.24920112e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-5.37177559e-01,  1.93979142e+00, -1.16971425e+00,
        -1.05217993e+00],
       [-1.50652052e+00,  7.88807586e-01, -1.34022653e+00,
        -1.18381211e+00],
       [-1.02184904e+00,  7.88807586e-01, -1.28338910e+00,
        -1.31544430e+00],
       [-1.74885626e+00, -3.62176246e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00,  9.82172869e-02, -1.28338910e+00,
        -1.44707648e+00],
       [-5.37177559e-01,  1.47939788e+00, -1.28338910e+00,
        -1.31544430e+00],
       [-1.26418478e+00,  7.88807586e-01, -1.22655167e+00,
      

In [21]:
std.fit_transform(data.target.reshape(-1,1))

array([[-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1.22474487],
       [-1

#### 1.1.2归一化/区间缩放（对列向量处理）

1. 把数据变成(0,1)或者(1,1)之间的小数。主要是为了数据处理方便提出来的，把数据映射到 0~1 范围之内处理，更加便捷快速。

2. 把有量纲表达式变成无量纲表达式，便于不同单位或量级的指标能够进行比较和加权。

$$x' = \frac{x-x_{min}}{x_{max}-x_{min}}$$

In [43]:
from sklearn.preprocessing import MinMaxScaler

In [44]:
minmax = MinMaxScaler()

In [45]:
minmax.fit_transform(data.data)

array([[0.22222222, 0.625     , 0.06779661, 0.04166667],
       [0.16666667, 0.41666667, 0.06779661, 0.04166667],
       [0.11111111, 0.5       , 0.05084746, 0.04166667],
       [0.08333333, 0.45833333, 0.08474576, 0.04166667],
       [0.19444444, 0.66666667, 0.06779661, 0.04166667],
       [0.30555556, 0.79166667, 0.11864407, 0.125     ],
       [0.08333333, 0.58333333, 0.06779661, 0.08333333],
       [0.19444444, 0.58333333, 0.08474576, 0.04166667],
       [0.02777778, 0.375     , 0.06779661, 0.04166667],
       [0.16666667, 0.45833333, 0.08474576, 0.        ],
       [0.30555556, 0.70833333, 0.08474576, 0.04166667],
       [0.13888889, 0.58333333, 0.10169492, 0.04166667],
       [0.13888889, 0.41666667, 0.06779661, 0.        ],
       [0.        , 0.41666667, 0.01694915, 0.        ],
       [0.41666667, 0.83333333, 0.03389831, 0.04166667],
       [0.38888889, 1.        , 0.08474576, 0.125     ],
       [0.30555556, 0.79166667, 0.05084746, 0.125     ],
       [0.22222222, 0.625     ,

**在什么时候使用标准化比较好，什么时候区间缩放比较好呢？**
1. 在后续的分类、聚类算法中，需要使用距离来度量相似性的时候、或者使用PCA、LDA这些需要用到协方差分析进行降维的时候，同时数据分布可以近似为正态分布，标准化方法(Z-score standardization)表现更好。（标准化好一点） 
2. 在不涉及距离度量、协方差计算、数据不符合正态分布的时候，可以使用区间缩放法或其他归一化方法。比如图像处理中，将RGB图像转换为灰度图像后将其值限定在[0 255]的范围。（归一化好点）

#### 1.1.3正则化（对行向量处理）

正则化的过程是将每个样本缩放到单位范数（每个样本的范数为1），如果后面要使用如二次型（点积）或者其它核方法计算两个样本之间的相似性这个方法会很有用。该方法主要应用于**文本分类和聚类**中。

主要思想是对每个样本计算其p-范数，然后对该样本中每个元素除以该范数，这样处理的结果是使得每个处理后样本的p-范数（L1-norm,L2-norm）等于1。

p-范数的计算公式：
$$||X||_p = \sqrt[p]{|x_1|^p+|x_2|^p+...+|x_n|^p}$$
规则为L2的归一化公式如下：
$$x' = \frac{x}{\sqrt{\sum^{m}_{j}{{x_j}^2}}}$$

In [46]:
from sklearn.preprocessing import Normalizer

In [47]:
norm = Normalizer()

In [48]:
norm.fit_transform(data.data)

array([[0.80377277, 0.55160877, 0.22064351, 0.0315205 ],
       [0.82813287, 0.50702013, 0.23660939, 0.03380134],
       [0.80533308, 0.54831188, 0.2227517 , 0.03426949],
       [0.80003025, 0.53915082, 0.26087943, 0.03478392],
       [0.790965  , 0.5694948 , 0.2214702 , 0.0316386 ],
       [0.78417499, 0.5663486 , 0.2468699 , 0.05808704],
       [0.78010936, 0.57660257, 0.23742459, 0.0508767 ],
       [0.80218492, 0.54548574, 0.24065548, 0.0320874 ],
       [0.80642366, 0.5315065 , 0.25658935, 0.03665562],
       [0.81803119, 0.51752994, 0.25041771, 0.01669451],
       [0.80373519, 0.55070744, 0.22325977, 0.02976797],
       [0.786991  , 0.55745196, 0.26233033, 0.03279129],
       [0.82307218, 0.51442011, 0.24006272, 0.01714734],
       [0.8025126 , 0.55989251, 0.20529392, 0.01866308],
       [0.81120865, 0.55945424, 0.16783627, 0.02797271],
       [0.77381111, 0.59732787, 0.2036345 , 0.05430253],
       [0.79428944, 0.57365349, 0.19121783, 0.05883625],
       [0.80327412, 0.55126656,

**StandardScaler和Normalizer这个两个概念的问题**

1. StandardScaler是尺寸缩放问题，即使同一特征下的数值在一定范围内浮动，如将数值所放在**0-1范围内(MinMaxScaler)**, 或者**将数据标准化成为均值为0，方差为1的数据(Z-score)**；
2. Normalizer，将同一行数据的不同特征进行规范化，这样一个数据的不同特征具有相同的量纲或者表现力，比如说一个特征是身高，1.7m，体重为150斤，那么两个特征之间差距太大，身高这个特征变化根本无法起到决定作用(在体重这个变化特征下)，毕竟大家怎么长都是一米多，但是体重差距一下子拉开20多是很正常的事。


### 1.2对定量特征二值化（对列向量）

**定性与定量区别**

定性：博主很胖，博主很瘦

定量：博主有80kg，博主有60kg 

定量特征二值化的核心在于设定一个阈值，大于阈值的赋值为1，小于等于阈值的赋值为0，公式表达如下：
$$x=
\begin{cases}
0& \text{x>threshold}\\
1& \text{x $\leq$ threshold}
\end{cases}$$


In [49]:
from sklearn.preprocessing import Binarizer

In [52]:
# 二值化，阈值设置为3，返回值为二值化后的数据
Binarizer(threshold=3).fit_transform(data.data)

array([[1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],


### 1.3对定性特征哑编码（对列向量处理）

one-hot编码

In [54]:
from sklearn.preprocessing import OneHotEncoder

In [109]:
ohenc = OneHotEncoder()

In [114]:
ohenc.fit_transform(data.target.reshape((-1,1)))

d:\b\python3.6\lib\site-packages\sklearn\preprocessing\_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


<150x3 sparse matrix of type '<class 'numpy.float64'>'
	with 150 stored elements in Compressed Sparse Row format>

In [125]:
onehot_target = ohenc.transform(data.target.reshape((-1,1))).toarray()
pd.DataFrame(onehot_target)

0    1    2
0    1.0  0.0  0.0
1    1.0  0.0  0.0
2    1.0  0.0  0.0
3    1.0  0.0  0.0
4    1.0  0.0  0.0
5    1.0  0.0  0.0
6    1.0  0.0  0.0
7    1.0  0.0  0.0
8    1.0  0.0  0.0
9    1.0  0.0  0.0
10   1.0  0.0  0.0
11   1.0  0.0  0.0
12   1.0  0.0  0.0
13   1.0  0.0  0.0
14   1.0  0.0  0.0
15   1.0  0.0  0.0
16   1.0  0.0  0.0
17   1.0  0.0  0.0
18   1.0  0.0  0.0
19   1.0  0.0  0.0
20   1.0  0.0  0.0
21   1.0  0.0  0.0
22   1.0  0.0  0.0
23   1.0  0.0  0.0
24   1.0  0.0  0.0
25   1.0  0.0  0.0
26   1.0  0.0  0.0
27   1.0  0.0  0.0
28   1.0  0.0  0.0
29   1.0  0.0  0.0
..   ...  ...  ...
120  0.0  0.0  1.0
121  0.0  0.0  1.0
122  0.0  0.0  1.0
123  0.0  0.0  1.0
124  0.0  0.0  1.0
125  0.0  0.0  1.0
126  0.0  0.0  1.0
127  0.0  0.0  1.0
128  0.0  0.0  1.0
129  0.0  0.0  1.0
130  0.0  0.0  1.0
131  0.0  0.0  1.0
132  0.0  0.0  1.0
133  0.0  0.0  1.0
134  0.0  0.0  1.0
135  0.0  0.0  1.0
136  0.0  0.0  1.0
137  0.0  0.0  1.0
138  0.0  0.0  1.0
139  0.0  0.0  1.0
140  0.0  0.0  1.0
141  0.0  0.0  1.0
142  0.0  0.0  1.0
143  0.0  0.0  1.0
144  0.0  0.0  1.0
145  0.0  0.0  1.0
146  0.0  0.0  1.0
147  0.0  0.0  1.0
148  0.0  0.0  1.0
149  0.0  0.0  1.0

[150 rows x 3 columns]

**为什么使用one-hot编码来处理离散型特征?**

在回归，分类，聚类等机器学习算法中，特征之间距离的计算或相似度的计算是非常重要的，而我们常用的距离或相似度的计算都是在欧式空间的相似度计算，计算余弦相似性，基于的就是欧式空间。

而我们使用one-hot编码，将离散特征的取值扩展到了欧式空间，离散特征的某个取值就对应欧式空间的某个点。

将离散型特征使用one-hot编码，确实会让特征之间的距离计算更加合理。

比如，有一个离散型特征，代表工作类型，该离散型特征，共有三个取值，不使用one-hot编码，其表示分别是x_1 = (1), x_2 = (2), x_3 = (3)。两个工作之间的距离是，(x_1, x_2) = 1, d(x_2, x_3) = 1, d(x_1, x_3) = 2。那么x_1和x_3工作之间就越不相似吗？显然这样的表示，计算出来的特征的距离是不合理。那如果使用one-hot编码，则得到x_1 = (1, 0, 0), x_2 = (0, 1, 0), x_3 = (0, 0, 1)，那么两个工作之间的距离就都是sqrt(2).即每两个工作之间的距离是一样的，显得更合理。

In [128]:
filepath = 'E:/#Mad-Lab/#DataSet/iris.csv'
iris_data = pd.read_csv(filepath,header=None,names=['sepal_length','sepal_width','petal_length','petal_width','target'])
iris_data.head()

sepal_length  sepal_width  petal_length  petal_width  target
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [134]:
category = pd.Categorical(iris_data.target)
code = category.codes
print(code)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [135]:
onehot_target2 = ohenc.transform(code.reshape((-1,1))).toarray()
pd.DataFrame(onehot_target2)

0    1    2
0    1.0  0.0  0.0
1    1.0  0.0  0.0
2    1.0  0.0  0.0
3    1.0  0.0  0.0
4    1.0  0.0  0.0
5    1.0  0.0  0.0
6    1.0  0.0  0.0
7    1.0  0.0  0.0
8    1.0  0.0  0.0
9    1.0  0.0  0.0
10   1.0  0.0  0.0
11   1.0  0.0  0.0
12   1.0  0.0  0.0
13   1.0  0.0  0.0
14   1.0  0.0  0.0
15   1.0  0.0  0.0
16   1.0  0.0  0.0
17   1.0  0.0  0.0
18   1.0  0.0  0.0
19   1.0  0.0  0.0
20   1.0  0.0  0.0
21   1.0  0.0  0.0
22   1.0  0.0  0.0
23   1.0  0.0  0.0
24   1.0  0.0  0.0
25   1.0  0.0  0.0
26   1.0  0.0  0.0
27   1.0  0.0  0.0
28   1.0  0.0  0.0
29   1.0  0.0  0.0
..   ...  ...  ...
120  0.0  0.0  1.0
121  0.0  0.0  1.0
122  0.0  0.0  1.0
123  0.0  0.0  1.0
124  0.0  0.0  1.0
125  0.0  0.0  1.0
126  0.0  0.0  1.0
127  0.0  0.0  1.0
128  0.0  0.0  1.0
129  0.0  0.0  1.0
130  0.0  0.0  1.0
131  0.0  0.0  1.0
132  0.0  0.0  1.0
133  0.0  0.0  1.0
134  0.0  0.0  1.0
135  0.0  0.0  1.0
136  0.0  0.0  1.0
137  0.0  0.0  1.0
138  0.0  0.0  1.0
139  0.0  0.0  1.0
140  0.0  0.0  1.0
141  0.0  0.0  1.0
142  0.0  0.0  1.0
143  0.0  0.0  1.0
144  0.0  0.0  1.0
145  0.0  0.0  1.0
146  0.0  0.0  1.0
147  0.0  0.0  1.0
148  0.0  0.0  1.0
149  0.0  0.0  1.0

[150 rows x 3 columns]

### 1.4 缺失值计算（对列向量处理）